dbms/src/Functions/array/range.cpp

```diff
@@ -27,79 +29,144 @@ class FunctionRange : public IFunction
 private:
     String getName() const override { return name; }
 
-    size_t getNumberOfArguments() const override { return 1; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
     bool useDefaultImplementationForConstants() const override { return true; }
 
     DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
     {
-        const DataTypePtr & arg = arguments.front();
-
-        if (!isUnsignedInteger(arg))
-            throw Exception{"Illegal type " + arg->getName() + " of argument of function " + getName(),
-                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+        if (arguments.size() > 3 || arguments.empty())
+        {
+            throw Exception{"Function " + getName() + " needs 1..3 arguments; passed "
+                            + std::to_string(arguments.size()) + ".",
+                            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH};
+        }
 
-        return std::make_shared<DataTypeArray>(arg);
+        for (const auto & arg : arguments)
+        {
+            if (!isUnsignedInteger(arg))
```

Comment:
But after this change it seems logical to also support Float, Decimal, Date, DateTime.

Reply:
And why we cannot have negative values for start, end, step?

Reply:
These changes can be postponed after merge.